app/src/main/java/de/triology/cas/services/RegistryLocal.java

```diff
@@ -217,31 +222,73 @@ public URI getCasLogoutUri(String doguname) throws GetCasLogoutUriException {
-                var previousServiceAccounts = readServiceAccounts();
-                while (true) {
-                    LOGGER.info("wait for changes under {}", LOCAL_CONFIG_DIR);
-                    watchService.take();
-                    var currentServiceAccounts = readServiceAccounts();
-                    if (!previousServiceAccounts.deepEquals(currentServiceAccounts)) {
-                        doguChangeListener.onChange();
-                        previousServiceAccounts = currentServiceAccounts;
+        Thread t1 = new Thread(() -> doguChangeListenerThreadHandler(doguChangeListener));
+        t1.start();
+    }
+
+    private void doguChangeListenerThreadHandler(DoguChangeListener doguChangeListener) {
+        var path = Paths.get(LOCAL_CONFIG_DIR);
+        WatchService watchService;
+        try {
+            watchService = initializeWatchService(path);
+            var previousServiceAccounts = readServiceAccounts();
+            WatchKey key;
+            LOGGER.info("wait for changes under {}", LOCAL_CONFIG_DIR);
+            while ((key = watchService.take()) != null) { // take does not return null
+                if (!key.isValid()) {
+                    LOGGER.info("watch key was cancelled or watch service was closed");
+                    LOGGER.info("try to reregister watch");
+                    watchService = initializeWatchService(path);
+                    continue;
+                }
+
+                List<WatchEvent<?>> watchEvents = key.pollEvents();
+                boolean serviceAccountRegistryChanged = false;
+                for (WatchEvent<?> event : watchEvents) {
+                    if (event.context().equals(LOCAL_CONFIG_FILE_NAME)) {
+                        serviceAccountRegistryChanged = true;
+                        break;
                     }
                 }
-            } catch (IOException e) {
-                throw new RegistryException("Failed to addDoguChangeListener", e);
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new RegistryException("Failed to addDoguChangeListener", e);
+
+                if (!serviceAccountRegistryChanged) {
+                    continue;
+                }
+
+                var currentServiceAccounts = readServiceAccounts();
+                if (!previousServiceAccounts.deepEquals(currentServiceAccounts)) {
+                    LOGGER.info("services changed. call doguChangeListener");
+                    doguChangeListener.onChange();
+                    previousServiceAccounts = currentServiceAccounts;
+                }
+
+                // Resetting the watchKey is very important. Otherwise, the key returned from take() (or poll()) will not return any more events.
+                if (!key.reset()) {
+                    LOGGER.info("reset key is no longer valid");
```

Comment:
reset key sounds like a key that is used for resetting, not a key that is being reset. I think "watch key" or something would fit better